### DISCUSSION TITLE
Begginer help with SWI

### DISCUSSION THREAD

**Discussion by hellenmir** [QSMxT/QSMxT#100] [Q&A] [answered]
Hello, 

Thanks a lot for this open-access pipeline!
I am very new to this field and I had a question on whether the pipeline can be run on UK-Biobank SWI imaging. I currently have 134 files as follows after running dcm2niix:
- SWI_TOTAL_MAG_notNorm_e1.nii.gz --> Total magnitude image - First echo time (1 file)
   - SWI_TOTAL_MAG_notNorm_e2.nii.gz --> Total magnitude image - Second echo time (1 file)
   - SWI_TOTAL_MAG_e1.nii.gz --> Total magnitude image (normalised by scanner) - First echo time (1 file)
   - SWI_TOTAL_MAG_e2.nii.gz --> Total magnitude image (normalised by scanner) - Second echo time (1 file)
   - SWI_TOTAL_PHASE_e1.nii.gz --> Total phase image - First echo time (1 file)
   - SWI_TOTAL_PHASE_e2.nii.gz --> Total phase image - Second echo time (1 file)
   - SWI_MAG_(c*)_e1.nii.gz --> Magnitude images per coil - First echo time(32 files)
   - SWI_MAG_(c*)_e2.nii.gz --> Magnitude images per coil - Second echo time (32 files)
   - SWI_PHASE_(c*)_e1.nii.gz --> Phase images per coil - First echo time(32 files)
   - SWI_PHASE_(c*)_e2.nii.gz --> Phase images per coil - Second echo time (32 files)
I will rename them as detailed in the BIDS extension protocol here: https://docs.google.com/document/d/1kyw9mGgacNqeMbp4xZet3RnDhcMmf4_BmRgKaOkO2Sc/edit 

But I was wondering if it possible to run the QSMxT pipeline on this data and if so how I would go about it. 

Thanks a lot for the help

**Comment by astewartau** [comment]
Hi @hellenmir, thanks for your interest in using QSMxT! I'd be pleased to help get this working for you.

I am unfamiliar with the biobank data at this stage, but if they are raw magnitude and phase images, you should be able to process them using QSMxT to produce both SWI and QSM (as well as T2* and R2* maps). You should only need the combined magnitude and phase images rather than the individual coil images. You will only need the individual coil images if the way they were combined did not preserve the phase correctly. 

It sounds like you have invoked `dcm2niix` manually, but I would recommend using QSMxT's `run_0_dicomSort.py` script followed by `run_1_dicomConvert.py` to convert to the Brain Imaging Data Structure (BIDS). You can convert to BIDS in other ways, but this should be the most straightforward. The extension proposal you cited is inactive - I'd recommend against following it at this stage because it's not yet part of the BIDS standard and won't be recognised by QSMxT. Also - you may need to delete the individual coil images before you run `run_1_dicomConvert.py` because QSMxT may not recognise these and might do strange things with them.. There isn't yet a standard way of storing individual coil data in BIDS, so the conversion behaviour is undefined. 

There's a nice example of using QSMxT from end-to-end here that I'd recommend as a quickstart:

https://www.neurodesk.org/example-notebooks/structural_imaging/qsmxt_example.html

You can run this notebook yourself at https://play.neurodesk.org/ and going into `example-notebooks` once you've started an instance of Neurodesk. 

Please let me know if you have any further questions or issues. I'd love to help you get it working for your data. :)

**Comment by marcellamontagnese** [comment]
Thanks so much for the quick reply! I have been working with Hellen on this and we managed to fix the issue by deleting the 32+32 individual coil images before running the `run_1_niftiConvert.py` as you suggested. This is a preferred option for us as we don't have access to all the DICOM files for all subjects.

It all seemed to work fine on the resulting images with multi-echo and we got these files as output:
sub-33940_ses-20220606_run-01_echo-01_part-mag_MEGRE.json
sub-33940_ses-20220606_run-01_echo-01_part-mag_MEGRE.nii.gz
sub-33940_ses-20220606_run-01_echo-01_part-phase_MEGRE.json
sub-33940_ses-20220606_run-01_echo-01_part-phase_MEGRE.nii.gz
sub-33940_ses-20220606_run-01_echo-02_part-mag_MEGRE.json
sub-33940_ses-20220606_run-01_echo-02_part-mag_MEGRE.nii.gz
sub-33940_ses-20220606_run-01_echo-02_part-phase_MEGRE.json
sub-33940_ses-20220606_run-01_echo-02_part-phase_MEGRE.nii.gz
sub-33940_ses-20220606_run-01_T1w.json
sub-33940_ses-20220606_run-01_T1w.nii.gz

I then managed to run the script for the analyses with the command 
`QSMxT/run_2_qsm.py 01_bids 02_qsm_output --do_qsm --do_swi --do_t2starmap --do_r2starmap --do_segmentation --auto_yes`
This runs correctly but  I am then having issues running the `run_5_analysis.py` script (I am skipping `run_4_template.py` for the testing for now). I see that the run_3_segmentations.py script has been incorporated into the previous script, so what would be the correct command now for running `run_5_analysis.py`?

Thanks a lot again for your help!

**Reply by astewartau** [reply]
Glad you got it converting nicely!

You are correct that on the current master branch, the `run_3_segmentations.py` file has been incorporated into the `run_2_qsm.py` script - this is a _very_ recent change, and you've correctly used `--do_segmentations` to run it. I will just note that by using the master branch you are getting these bleeding-edge changes and some instability is to be expected, though you might be lucky and everything runs smoothly for your analyses. :crossed_fingers: 

The `run_4_template.py` script is fine to skip and unnecessary for the later one.

The `run_5_analysis.py` script requires:
 - a list of segmentations; e.g. `--segmentations run_2_output/segmentations/qsm_segmentations/*.nii*`;
 - a list of QSMs; e.g. `--qsm_files run_2_output/qsm/*.nii`;
 - optionally, a CSV file that maps the integer labels from the segmentations to named regions - for the segmentations generated by QSMxT, this is the `aseg_labels.csv` file in the repository or in `/opt/QSMxT/aseg_labels.csv` in the container
 
Importantly, the segmentations and QSM files should match pairwise under the same alphanumeric sorting, which should be true if they were all generated using QSMxT. 

Hopefully, that helps! Happy to help with any further issues.

**Comment by marcellamontagnese** [comment]
Great thank you!
I have updated the scripts to your latest edits and run `run_2_qsm.py` with the `fast` preset and specifying `--do_qsm --do_swi --do_t2starmap --do_r2starmap --do_segmentation --auto_yes`
I am attaching the tree directory structure of the resulting 02_qsm_output folder. [tree_02_qsm_output.txt](https://github.com/QSMxT/QSMxT/files/12300873/tree_02_qsm_output.txt)

I then tried running `/home/mm2075/rds/hpc-work/.conda/envs/qsmxt/bin/python QSMxT/run_5_analysis.py --segmentations 02_qsm_output/segmentations/*.nii* --qsm_files 02_qsm_output/qsm/*/*.nii --output_dir analysis --labels_file QSMxT/aseg_labels.csv` based on the 02_qsm_output fodler structure

But here I keep getting the error `ValueError: operands could not be broadcast together with shapes (3538944,) (16777216,) `
I am atttaching the full terminal output in case it is of more help:


`/home/mm2075/rds/hpc-work/.conda/envs/qsmxt/bin/python QSMxT/run_5_analysis.py --segmentations 02_qsm_output/segmentations/*.nii* --qsm_files 02_qsm_output/qsm/*/*.nii --output_dir analysis --labels_file QSMxT/aseg_labels.csv
[INFO]: Running QSMxT v3.2.1-50-gb91af61 (commit date: Wed Aug 9 15:25:51 2023 +1000)
[INFO]: Command: QSMxT/run_5_analysis.py --segmentations 02_qsm_output/segmentations/sub-33909_ses-20220505_run-01_T1w_segmentation_nii.nii 02_qsm_output/segmentations/sub-33909_ses-20220505_run-01_T1w_segmentation_nii_trans.nii 02_qsm_output/segmentations/sub-33909_ses-20220505_run-02_T1w_segmentation_nii.nii 02_qsm_output/segmentations/sub-33909_ses-20220505_run-02_T1w_segmentation_nii_trans.nii 02_qsm_output/segmentations/sub-33916_ses-20220518_run-01_T1w_segmentation_nii.nii 02_qsm_output/segmentations/sub-33916_ses-20220518_run-01_T1w_segmentation_nii_trans.nii 02_qsm_output/segmentations/sub-33916_ses-20220518_run-02_T1w_segmentation_nii.nii 02_qsm_output/segmentations/sub-33916_ses-20220518_run-02_T1w_segmentation_nii_trans.nii 02_qsm_output/segmentations/sub-33939_ses-20220606_run-01_T1w_segmentation_nii.nii 02_qsm_output/segmentations/sub-33939_ses-20220606_run-01_T1w_segmentation_nii_trans.nii 02_qsm_output/segmentations/sub-33939_ses-20220606_run-02_T1w_segmentation_nii.nii 02_qsm_output/segmentations/sub-33939_ses-20220606_run-02_T1w_segmentation_nii_trans.nii 02_qsm_output/segmentations/sub-33940_ses-20220606_run-01_T1w_segmentation_nii.nii 02_qsm_output/segmentations/sub-33940_ses-20220606_run-01_T1w_segmentation_nii_trans.nii 02_qsm_output/segmentations/sub-33940_ses-20220606_run-02_T1w_segmentation_nii.nii 02_qsm_output/segmentations/sub-33940_ses-20220606_run-02_T1w_segmentation_nii_trans.nii --qsm_files 02_qsm_output/qsm/_qsmjl_rts0/sub-33909_ses-20220505_run-01_echo-01_part-phase_MEGRE_scaled_romeo-b0map_normalized_vsharp_rts.nii 02_qsm_output/qsm/_qsmjl_rts0/sub-33916_ses-20220518_run-01_echo-01_part-phase_MEGRE_scaled_romeo-b0map_normalized_vsharp_rts.nii 02_qsm_output/qsm/_qsmjl_rts0/sub-33939_ses-20220606_run-01_echo-01_part-phase_MEGRE_scaled_romeo-b0map_normalized_vsharp_rts.nii 02_qsm_output/qsm/_qsmjl_rts0/sub-33940_ses-20220606_run-01_echo-01_part-phase_MEGRE_scaled_romeo-b0map_normalized_vsharp_rts.nii --output_dir analysis --labels_file QSMxT/aseg_labels.csv`

`[INFO]: Python interpreter: /home/mm2075/rds/hpc-work/.conda/envs/qsmxt/bin/python
[INFO]: Analysing file sub-33909_ses-20220505_run-01_echo-01_part-phase_MEGRE_scaled_romeo-b0map_normalized_vsharp_rts.nii with segmentation sub-33909_ses-20220505_run-01_T1w_segmentation_nii.nii`

`Traceback (most recent call last):
  File "QSMxT/run_5_analysis.py", line 332, in <module>
    run_analysis(args)
  File "QSMxT/run_5_analysis.py", line 324, in run_analysis
    calculate_statistics(args, logger)
  File "QSMxT/run_5_analysis.py", line 258, in calculate_statistics
    one_segmentation_per_subject(args, logger)
  File "QSMxT/run_5_analysis.py", line 206, in one_segmentation_per_subject
    label_stats = get_stats(labels, seg, qsm)
  File "QSMxT/run_5_analysis.py", line 58, in get_stats
    qsm_values = qsm[np.logical_and(qsm != 0, qsm_seg)]
ValueError: operands could not be broadcast together with shapes (3538944,) (16777216,)`

**Reply by astewartau** [reply, marked as answer]
Thanks for those details!

This is happening because the segmentations folder includes both segmentations in the T1w space and the QSM space - I intend to clean up the output a bit to make this more apparent by putting them into their own folder.

You can fix this by using `--segmentations 02_qsm_output/segmentations/*_trans.nii*` - this should only match the segmentations files that are in the QSM space (the `_trans` suffix indicates transformed) and should therefore have the same dimensions and be able to be 'broadcast together' as per the error message.

**Reply by marcellamontagnese** [reply]
Great. Thank you very much for the assistance with this!

**Reply by marcellamontagnese** [reply]
Sorry I just had a follow-up question. My output csv files have around 96 regions with information in them and they seem to include several voxels each. Is it possible to get the value for each voxel as part of the output?

**Reply by astewartau** [reply]
No problem! The CSV files provide summary statistics of susceptibility values across those 96 regions, including the minimum, maximum, mean and standard deviation in ppm. Since most regions contain many voxels, the CSV can't practically store information for each voxel. For these individual values, I'd suggest deriving them from the QSM volumes directly, along with the relevant segmentation. The example notebook shows how to do this in Python to generate some basic boxplots for a few different regions - https://www.neurodesk.org/example-notebooks/structural_imaging/qsmxt_example.html#analysis-csvs

**Reply by marcellamontagnese** [reply]
That's perfect! Thanks once more for the prompt and very helpful reply
